### PR TITLE
Display progress for mailinglist send

### PR DESCRIPTION
**Pull request by RoukePouw**
#332

Extend makeAction function with javascript callback
Set intermediate progress updates in xml
Lock and unload browser page with javascript.

### CODE REVIEW

File: system/datatypes/mailinglist.php [hunk 1]
```diff
@@ -702,7 +702,13 @@ protected function mailingView(HyphaRequest $request) {
 			$num = count($addresses);
 
 			$linkToMailing = $this->path(self::PATH_MAILS_VIEW_ID, ['id' => $mailingId]);
-			$action = 'if(confirm(' . json_encode(__('ml-sure-to-send', ['count' => $num])) . '))' . makeAction($linkToMailing, self::CMD_SEND, '');
+
+			$action = 'if(confirm(' . json_encode(__('ml-sure-to-send', ['count' => $num])) . ')){
+				const checkProgress = ()=>{}; // TODO do a call to get the progress and display this in a bar
+				setInterval(checkProgress,1000); //TODO make into a constant
+				window.onbeforeunload = ()=>\'Are you sure you want to leave?\';' // lock the browser from exiting. TODO translation?
```
Comment (matthijskooijman) [inline system/datatypes/mailinglist.php:709]: If this is shown to the user, it should probably be more explicit about what happens when you leave?

**Review by matthijskooijman — COMMENTED** [review]
I had a quick look at this code, looks like it's heading in a good direction. Before the final merge, I think some things (at least the callback argument to `makeAction`) should be moved into separate commits, of course.

Also, this progress tracking could maybe also result in resumable mailinglist sending, I think (because subscribers are never deleted, only changes status when they unsubscribe IIRC, so the `progress` attribute can be used to pick up where an interrupted sending left off). Probably not in scope for this PR, but good to keep in mind :-)

**Comment by matthijskooijman** [comment]
Also, be sure to put "Fixes: #332" in the commit message that actually implements the core of this PR, so that issue can be autoclosed.

File: system/datatypes/mailinglist.php [hunk 2]
```diff
@@ -697,14 +704,60 @@ protected function mailingView(HyphaRequest $request) {
...
+			  const checkProgress = function(){'. // retrieve the progress of the send action
+					makeAction($linkToMailing, self::CMD_PROGRESS, '', null, 'updateProgress')
+				.'};
+				checkProgressInterval = setInterval(checkProgress,1000); //TODO make into a constant
+				window.onbeforeunload = function(){return \'Are you sure you want to leave?\';};' // lock the browser from exiting. TODO translation?
+				 . makeAction($linkToMailing, self::CMD_SEND, '', null, 'onCompleted').';
+			 }';
+
+			if ($num > 0){ // hide send button if no addresses available
+				$action = str_replace(['"'], ['\''], $action);
```
Comment (matthijskooijman) [inline system/datatypes/mailinglist.php:757]: Why is this replace needed? If this is because `"` breaks the `onclick="..."` quoting, then maybe this should be fixed by using `htmlspecialchars` in `makeAction` instead.

Comment (matthijskooijman) [reply system/datatypes/mailinglist.php:757]: Seems this is from existing code, but still interesting to look into this.

File: system/datatypes/mailinglist.php [hunk 3]
```diff
@@ -1018,18 +1073,31 @@ protected function mailingSendAction(HyphaRequest $request) {
...
 			$linkToUnsubscribe = $this->path(self::PATH_UNSUBSCRIBE_CODE, ['address' => $email, 'code' => $code]);
 			$message = '<p><a href="' . $linkToMailing . '">' . __('ml-if-unreadable-use-link') . '</a></p>';
 			$message .= $mailing->getHtml();
 			$message .= '<p><a href="' . $linkToUnsubscribe . '">' . __('ml-unsubscribe') . '</a></p>';
-			$this->sendMail($mailing->getAttribute('subject'), $message, [$email], $senderData['email'], $senderData['name']);
+			//TODO disabled for testing purposes $this->sendMail($mailing->getAttribute('subject'), $message, [$email], $senderData['email'], $senderData['name']);
+
+			$this->xml->lockAndReload();
+			$mailing = $this->xml->getElementById($mailingId);
+			$mailing->setAttribute('progress', ++$count);
```
Comment (matthijskooijman) [inline system/datatypes/mailinglist.php:1097]: Should the progress attribute be removed when the send is complete?

File: system/datatypes/mailinglist.php [hunk 4]
```diff
@@ -1083,6 +1151,25 @@ protected function mailingSendTestAction(HyphaRequest $request) {
...
+
+		$mailingId = $request->getArg(1);
+
+		// check if given id was correct
+		/** @var HyphaDomElement $mailing */
+		$mailing = $this->xml->getElementById($mailingId);
+		if (!$mailing instanceof HyphaDomElement) {
+			return '404';
+		}
+		$progress = $mailing->getAttribute('progress').'/'.$this->getNumberOfAddresses();
```
Comment (matthijskooijman) [inline system/datatypes/mailinglist.php:1167]: Would it be better to return JSON instead of a 0/0 format that needs to be split client-side? JSON is easier and more robust to parse?

File: system/datatypes/mailinglist.php [hunk 5]
```diff
@@ -1083,6 +1151,25 @@ protected function mailingSendTestAction(HyphaRequest $request) {
 		return 'reload';
 	}
 
+	protected function mailingProgressAction(HyphaRequest $request) {
+		if (!$this->hasSender()) {
```
Comment (matthijskooijman) [inline system/datatypes/mailinglist.php:1154]: Is this check really relevant here? Seems like a copy-paste that is not really needed here?

**Review by matthijskooijman — COMMENTED** [review]
Looks good. Left some comments inline (I think you already fixed translations, so I'll resolve those comments directly if they're fixed).

Overall:
 - This should be split over multiple commits.
 - There is a *lot* of inline JS. I wonder if that would be better in a script tag, which can be done using the HTMLDocument `writeScript()` method.

File: system/languages/en.php [hunk 6]
```diff
@@ -313,6 +313,11 @@
 		"ml-received-by" => "received by",
 		"ml-sure-to-send" => "Are you sure you want to send off this mailing to [[count]] recipient(s)?",
 
+		// mailing list - progress feedback ui
+		"ml-sure-to-close" => "Are you sure you want to leave and stop sending the mailing?",
+		"ml-do-not-close" => "Attention: Do not close this window untill all mails are sent.",
+		"ml-send-in-progress" => "The mail is being send...",
```
Comment (matthijskooijman) [inline system/languages/en.php:319]:
```suggestion
		"ml-send-in-progress" => "The mail is being sent...",
```

File: system/datatypes/mailinglist.php [hunk 7]
```diff
@@ -697,14 +704,60 @@ protected function mailingView(HyphaRequest $request) {
...
-			$num = count($addresses);
-
+			$num = $this->getNumberOfAddresses();
 			$linkToMailing = $this->path(self::PATH_MAILS_VIEW_ID, ['id' => $mailingId]);
-			$action = 'if(confirm(' . json_encode(__('ml-sure-to-send', ['count' => $num])) . '))' . makeAction($linkToMailing, self::CMD_SEND, '');
-			$action = str_replace(['"'], ['\''], $action);
-			$commands->append(makeButton(__('send'), $action));
+
+			$action = 'if(confirm(' . json_encode(__('ml-sure-to-send', ['count' => $num])) . ')){
+				const divPopUp = document.createElement("DIV"); // create a popup
```
[the remark this answers is not in the document]
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:712]: Fixed :)

File: system/datatypes/mailinglist.php [hunk 8]
```diff
@@ -697,14 +704,60 @@ protected function mailingView(HyphaRequest $request) {
...
+				spanProgressIndicator.innerHTML = "0/'.$num.'";
+				const divWarning = document.createElement("div");
+				divWarning.className = "warning";
+				divWarning.innerHTML = "Let op: Sluit dit venster niet tot de hele mailing is verstuurd";
+				divProgress.appendChild(spanProgressText);
+				divProgress.appendChild(spanProgressIndicator);
+				divPopUp.appendChild(divProgress);
+				divPopUp.appendChild(divWarning);
+
+				const divGray = document.createElement(\'DIV\'); // create a gray layer
```
[the remark this answers is not in the document]
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:731]: Fixed :)

File: system/datatypes/mailinglist.php [hunk 9]
```diff
@@ -697,14 +704,60 @@ protected function mailingView(HyphaRequest $request) {
...
+					window.onbeforeunload = undefined;
+				};
+				const updateProgress = function(progress){
+					const [done,total] = progress.split("/").map(Number);
+					if(done>=total) onCompleted();
+					else{
+						spanProgressIndicator.innerHTML = progress;
+					}
+				};
+			  const checkProgress = function(){'. // retrieve the progress of the send action
```
[the remark this answers is not in the document]
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:748]: Fixed :)

File: system/datatypes/mailinglist.php (hunk 3 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:1097]: Then the progress checker will find no final progress '10/10' and cannot determine that it has finished (using the way it has been build now).

File: system/datatypes/mailinglist.php (hunk 4 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:1167]: We can do that. Better/easier is a bit subjective  :-) 

JSON is used because is it both human readable and computer parsable. If find `'5/10'` more readable than `{"current":5,"total":10}`. 
Parsing is `const {current,total} = JSON.parse(string);` vs `const [current,total] = string.split('/');` not much difference in complexity there.

And as to robustness `(x,y) => x+'\'+y` is a fully bijective function as for x,y without slashes so parsing is as robust as it gets (and probably even faster than JSON parse)

But I'm more than fine to switch to JSON :-)

File: system/datatypes/mailinglist.php (hunk 2 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:757]: Yeah, I just kept it there. I can look into it if that's helpful?

File: system/languages/en.php (hunk 6 above)
Comment (RoukePouw) [reply system/languages/en.php:319]: Fixed :)

File: system/datatypes/mailinglist.php (hunk 5 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:1154]: A copy paste  indeed. I kept it as it seemed to be used in other functions. It made sense to me that a mailing without a sender would fail to return progress. 
Should I remove it?

**Comment by RoukePouw** [comment]
> There is a lot of inline JS. I wonder if that would be better in a script tag, which can be done using the HTMLDocument writeScript() method.

Now sure how that would work, can you explain?

In other cases I would define a file with the functions that would get packed together into a minified library or something. But with hypha it seems to be the convention to group code together in a single file.

**Comment by RoukePouw** [comment]
> This should be split over multiple commits.

How would you like this done?

File: system/datatypes/mailinglist.php (hunk 4 above)
Comment (matthijskooijman) [reply system/datatypes/mailinglist.php:1167]: Agreed, but JSON is also more extendable (and if we add other AJAX APIs, can be used more consistently), so I'd still favor JSON here.

File: system/datatypes/mailinglist.php (hunk 3 above)
Comment (matthijskooijman) [reply system/datatypes/mailinglist.php:1097]: But the progress endpoint could just look at the status for that? Even more, if the endpoint is switched to JSON, it could maybe even return a "status" (or "completed") attributed as well?

File: system/datatypes/mailinglist.php (hunk 2 above)
Comment (matthijskooijman) [reply system/datatypes/mailinglist.php:757]: I suspect there is some fragility in the current system because of this, so if you want to have look, please do.

File: system/datatypes/mailinglist.php (hunk 4 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:1167]: Agreed :)

File: system/datatypes/mailinglist.php (hunk 5 above)
Comment (matthijskooijman) [reply system/datatypes/mailinglist.php:1154]: I would remove it for simplicity.

File: system/datatypes/mailinglist.php (hunk 3 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:1097]: Yes, will do that :)

**Comment by matthijskooijman** [comment]
> How would you like this done?

Do you mean "Through what process"? Or "split into which commits"?

**Comment by RoukePouw** [comment]
> Do you mean "Through what process"? Or "split into which commits"?

First in which commits you would like it split. (and why :-) so I can understand the thoughts behind it ) Secondly, yes, if you have pointers on how to do this, otherwise I'll do some DuckDuckGo-ing myself ;-)

File: system/datatypes/mailinglist.php (hunk 5 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php]: Done :)

File: system/datatypes/mailinglist.php (hunk 4 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php]: And done

File: system/datatypes/mailinglist.php (hunk 3 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:1099]: And done

File: system/datatypes/mailinglist.php (hunk 2 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:759]:
```
system/core/various.php:98

function makeButton($label, $action, $id = '', $class = '') {
		$class = 'button' . ($class ? ' ' . $class : '');
		return '<input type="button" class="'.$class.'" '.($id?' id="'.$id.'"':'').'value="'.$label.'" onclick="'.$action.'" />';
	}
```

I would suggest doing a fix here instead of in mailinglist.php. We could do something along these lines:

```
function makeButton($label, $action, $id = '', $class = '') {
		$class = 'button' . ($class ? ' ' . $class : '');
		$action = makeSave($action);
		return '<input type="button" class="'.$class.'" '.($id?' id="'.$id.'"':'').'value="'.$label.'" onclick="'.$action.'" />';
	}
```

but I wouldn't recommend using `htmlspecialchars`. Because that will make html entities out of javascript code, right?
Simply replacing `"` with `\'` could also lead to unpredictable behavior, depending on your javascript usage of quotes. Same with replacing `"` with `\"`. There's always an edge case that could be triggered.

Best way I can think of might be:

```
function makeButton($label, $action, $id = '', $class = '') {
		$class = 'button' . ($class ? ' ' . $class : '');
		$functionName = '?';
		return '<script>function '.$functionName.'(){'.$action.'}</script><input type="button" class="'.$class.'" '.($id?' id="'.$id.'"':'').'value="'.$label.'" onclick="'$functionName'();" />';
	}
```

But then we would need a proper way of determining a good unique name for `$functionName`;

**Comment by matthijskooijman** [comment]
I would make commits to:
 - Remove the extra request headers
 - Support callback for `makeAction`
 - Fix for `str_replace(['"'], ['\''], $action);` if you find it
 - Optionally the CSS for the messagebox, because it is fairly general (but probably better to just include this in the next commit)
 - Everything else (i.e. implementing the progress counter)

As for the how: In this case I think none of the current commits are usable as is, so it is probably best to do a (soft) reset back to the original commit (just before your first commit), to effectively remove all commits, but leave the changes in your working copy, so you can use e.g. `git add -p` to create new commits from that. If you would want to preserve existing commits but split or rearrange some, then using an interactive rebase would be the way to go.

File: system/datatypes/mailinglist.php (hunk 2 above)
Comment (matthijskooijman) [reply system/datatypes/mailinglist.php:759]: > but I wouldn't recommend using htmlspecialchars. Because that will make html entities out of javascript code, right?

It would, but I don't think that's a problem. Since HTML attributes  (such as onclick) should be entity-decoded before they are used (this should be handled transparently by the DOM parser), so the javascript interpreter should see exactly the input the `htmlspecialchars`. One notable exception is CDATA (inside CDATA blocks or some tags), which is used literal, but that doesn't apply here AFAICS.

> Best way I can think of might be:

Clever approach, but it also seems a bit fragile. And indeed, it needs unique function names.

Comment (RoukePouw) [reply system/datatypes/mailinglist.php:759]: htmlspecialchars it is :-)

**Comment by RoukePouw** [comment]
Commits have been split :)

File: system/datatypes/mailinglist.php (hunk 2 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php]: Though this does leave the edge case that, for example, `console.log('&gt;')` will not perform as expected.

Comment (matthijskooijman) [reply system/datatypes/mailinglist.php]: Why not? Did you test this? I think this should be encoded as `console.log(&amp;gt;')` in the HTML, which should be decoded into the proper code again before running the javascript?

Comment (RoukePouw) [reply system/datatypes/mailinglist.php]: My bad, apparently I tested it to hastily and messed up enabling the `htmlspecialchars` :-S

File: system/datatypes/mailinglist.php [hunk 10]
```diff
@@ -1083,6 +1149,24 @@ protected function mailingSendTestAction(HyphaRequest $request) {
...
+		/** @var HyphaDomElement $mailing */
+		$mailing = $this->xml->getElementById($mailingId);
+		if (!$mailing instanceof HyphaDomElement) {
+			return '404';
+		}
+		$status = $mailing->getAttribute(self::FIELD_NAME_STATUS);
+		$progress = $mailing->getAttribute(self::FIELD_NAME_PROGRESS);
+		$total = $this->getNumberOfAddresses();
+		$result = [ self::FIELD_NAME_PROGRESS => $progress, self::FIELD_NAME_TOTAL => $total, self::FIELD_NAME_STATUS => $status];
+		echo json_encode($result);
```
Comment (matthijskooijman) [inline system/datatypes/mailinglist.php:1166]: This should probably also set a Content-type header, to allow browsers and client frameworks to know that his is JSON. Not strictly required now, but probably helpful for debugging nonetheless.

File: system/datatypes/mailinglist.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (matthijskooijman) [reply system/datatypes/mailinglist.php:750]: Nope, seems to still be there. The file is indented with tabs, and this line seems use tabs followed by one level of 2 spaces (which doesn't show up when you display tabs as 2 spaces, but is inconsistent and shows up with other tab stop settings).

File: system/datatypes/mailinglist.php [hunk 11]
```diff
@@ -1083,6 +1149,24 @@ protected function mailingSendTestAction(HyphaRequest $request) {
 		return 'reload';
 	}
 
+	protected function mailingProgressAction(HyphaRequest $request) {
+
```
Comment (matthijskooijman) [inline system/datatypes/mailinglist.php:1153]: Extra empty line.

File: system/datatypes/mailinglist.php [hunk 12]
```diff
@@ -1010,6 +1062,8 @@ protected function mailingSendAction(HyphaRequest $request) {
 		$this->xml->lockAndReload();
 		$mailing = $this->xml->getElementById($mailingId);
 		$mailing->setAttribute('status', self::MAILING_STATUS_SENDING);
+		$count = 0;
+		$mailing->setAttribute('progress', $count);
```
Comment (matthijskooijman) [inline system/datatypes/mailinglist.php:1066]: This should use a constant. Same for `status` btw, and some other string constants in this function, but those would be a separate comit (would be good if you could fix them up, though).

**Review by matthijskooijman — COMMENTED** [review]
I left some nitpick comments, overall it looks great. I also tested, worked like a charm here :-)

Regarding commit structure:
 - I would put commit "Add css changes for mailing progress ui" directly after "Display progress for mailinglist send" since it is tightly coupled.
 - The commit message for "Display progress for mailinglist send" still has some old points included and is an enumeration of points which aren't terribly clear on their own. Could you update that to a self-standing commit message (good to name the subparts of the commit, but maybe with a few more words).

File: system/core/various.php [hunk 13]
```diff
@@ -97,7 +97,7 @@ function asterisk($condition) {
 	*/
 	function makeButton($label, $action, $id = '', $class = '') {
 		$class = 'button' . ($class ? ' ' . $class : '');
-		return '<input type="button" class="'.$class.'" '.($id?' id="'.$id.'"':'').'value="'.$label.'" onclick="'.$action.'" />';
+		return '<input type="button" class="'.$class.'" '.($id?' id="'.$id.'"':'').'value="'.$label.'" onclick="'.htmlspecialchars($action).'" />';
```
[the remark this answers is not in the document]
Comment (RoukePouw) [reply system/core/various.php]: Done :)

File: system/datatypes/mailinglist.php (hunk 10 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php:1166]: Done :)

File: system/datatypes/mailinglist.php (hunk 11 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php]: Done :)

File: system/datatypes/mailinglist.php (hunk 12 above)
Comment (RoukePouw) [reply system/datatypes/mailinglist.php]: Done :)

**Comment by matthijskooijman** [comment]
I made a few more changes:
 - Added "mailinglist: " prefix to commit message
 - Converted a few more `'progress'` to `self::FIELD_NAME_PROGRESS`
 - Converted a few other strings to constants too (as a separate commit) and moved one conversion of `'status'` to this new commit.
 - Fixed one incorrect type annotation I noticed (was already broken in master).
 - Rebased on top of master.

 With that, I'll merge this.

**Review by matthijskooijman — APPROVED** [review]
